### ISSUE TITLE
Is it possible to accumulate gradients over multiple batches?

### ISSUE THREAD

**Issue by nicklhy** [apache/incubator-mxnet#3283]
Due to the limit of GPU memory, I can not use a large mini-batch size when training a deep network like ResNet or Inception-v3. But I noticed that there is a solution in caffe that use the parameter 'iter_size' to accumulate gradients over multiple batches(i.e. accumulate gradients 10 times and then update the weights). So, is there a similar parameter setting in mxnet? If not, will it be easy to implement it myself?

**Comment by tornadomeet** [comment]
i think there is no iter like in caffe.
but you can try https://github.com/dmlc/mxnet-memonger

**Comment by nicklhy** [comment]
@tornadomeet , thanks for your reply but the memory cost is actually the same for my training code after using the "memonger" like
```
net = mx.sym.load(net_json)
net = train.search_plan(net, data=(batch_size, 3, height, width))

model = mx.FeedForward(net_planned, ...)
model.fit(...)
```
Moreover, I think the "memonger" is still far from what I really need(i.e. to reduce the memory to only 1/10 or 1/20). If the iter_size is not implemented yet in mxnet, I should manage to do it myself.

**Comment by tornadomeet** [comment]
the committer of `memonger` say it can training resnet-1000 on imanget using batchsize=32 on a single gpu(<8G memory), but i'm not sure whether it can train incpetion-v3 of 10X batch size. you can hack into it.

**Comment by nicklhy** [comment]
@tornadomeet , it seems that my previous result is not good because I did not set attribute mirror_stage='True' in Inception-v3 properly. I tried the ResNet-50 just now exactly as the example code and found the GPU memory dropped from 5555MiB to 2861MiB(batch-size=40). Thus, I should be able to change batch-size double in my GTX 1080 now.

But from my own perspective, iter_size in caffe could still be a more natural way to balance the GPU memory limit and mini-batch size. We can easily reduce 90% GPU memory just by setting iter_size to 10. We don't need to think about which layers could possibly be the places that need to bookkeep computations. And what's more, accumulating gradients over multiple batches is also much easier to implement.

**Comment by tornadomeet** [comment]
yes, but be careful with 'iter', 'iter' may bring worse result when using bn layer.

**Comment by nicklhy** [comment]
@tornadomeet , do you mean the mean and variance computed from N different mini-batch data(iter_size=N) will lead to different results comparing with the values from a single N*batch_size batch data?

**Comment by tornadomeet** [comment]
yes

**Comment by githubfragments** [comment]
I also want to use 'iter_size' to accumulate gradients over multiple batches. Caffe has a special parameter 'iter_size' to accumulate. Pytorch also can realize the function by  not zeroing the grad to keep accumulating the gradient in. How to  realize in mxnet? @tqchen

**Comment by tqchen** [comment]
It is possible with low level api, bind gradient with add_to req, and no zeroing the gradient over iterations

**Comment by githubfragments** [comment]
@tqchen You mean just change grad_req to 'add', and update every iter_size iterations. The gradients will auto zeroing every update(). like this:
mod.bind(data_shapes=.., label_shapes=.., grad_req='add')
mod.init_params(..)
mod.init_optimizer(..)
for epoch in range(total_epochs):
    cnt = 0
    for batch in train_iter:
        cnt += 1
        mod.forward_backward(batch)  # compute predictions
        if cnt % iter_size == 0:    # every iter_size iterations 
              mod.update()  # update parameters

**Comment by tqchen** [comment]
I am not sure if mod.update will zeros the gradient, so maybe gradient zero need to be done manually

**Comment by githubfragments** [comment]
Oh. How to manually zero the gradient after each update, is there a simple example?

**Comment by realwecan** [comment]
@githubfragments I'm also interested in this feature. Could you share with us if you have been able to do this? Thanks!

**Comment by githubfragments** [comment]
Sorry, I didn't keep working this out.@realwecan

**Comment by githubfragments** [comment]
Do you figure out how to implement accumulate gradients over multiple batches? @realwecan @nicklhy

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
